File: android/src/main/java/com/reactlibrary/McuManagerModule.java
package com.reactlibrary;

import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.net.Uri;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableMap;

import com.facebook.react.modules.core.DeviceEventManagerModule;

public class McuManagerModule extends ReactContextBaseJavaModule {

    private final ReactApplicationContext reactContext;
    private final BluetoothAdapter bluetoothAdapter;
    DeviceUpdate update;

    public McuManagerModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;

        this.bluetoothAdapter = BluetoothAdapter.getDefaultAdapter();
    }

    @Override
    public String getName() {
        return "McuManager";
    }

    @ReactMethod
    public void updateDevice(String macAddress, String updateFileUriString, Promise promise) {
        if this.update == null {
            promise.reject("an update is already running");
        }
        BluetoothDevice device = bluetoothAdapter.getRemoteDevice(macAddress);
        Uri updateFileUri = Uri.parse(updateFileUriString);

        this.update = new DeviceUpdate(device, promise, reactContext, updateFileUri, this);
        this.update.startUpdate();
    }

    @ReactMethod
    public void cancelRunningUpdates() {
        if this.update != null {
            this.update.cancel();
        }
    }

    public void updateProgressCB(WritableMap progress) {
        this.reactContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
            .emit("uploadProgress", progress);
    }

    public void updateStateCB(WritableMap state) {
        this.reactContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
            .emit("uploadStateChanged", state);
    }

    public void unsetUpdate() {
        this.update = null;
    }
}

File: android/src/main/java/uk/co/playerdata/reactnativemcumanager/McuManagerModule.kt
package uk.co.playerdata.reactnativemcumanager

import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.net.Uri
import android.util.Log
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule.RCTDeviceEventEmitter
import io.runtime.mcumgr.ble.McuMgrBleTransport
import io.runtime.mcumgr.managers.ImageManager

class McuManagerModule(private val reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
    private val TAG = "McuManagerModule"
    private val bluetoothAdapter: BluetoothAdapter? = BluetoothAdapter.getDefaultAdapter()
    private val upgrades: MutableMap<String, DeviceUpgrade> = mutableMapOf()

    override fun getName(): String {
        return "McuManager"
    }

    @ReactMethod
    fun eraseImage(macAddress: String?, promise: Promise) {
        if (this.bluetoothAdapter == null) {
            throw Exception("No bluetooth adapter")
        }

        try {
            val device: BluetoothDevice = bluetoothAdapter.getRemoteDevice(macAddress)

            var transport = McuMgrBleTransport(reactContext, device)
            transport.connect(device).timeout(60000).await()

            val imageManager = ImageManager(transport);
            imageManager.erase()

            promise.resolve(null)
        } catch (e: Throwable) {
            promise.reject(e)
        }
    }

    @ReactMethod
    fun createUpgrade(id: String, macAddress: String?, updateFileUriString: String?, updateOptions: ReadableMap) {
        if (this.bluetoothAdapter == null) {
            throw Exception("No bluetooth adapter")
        }

        if (upgrades.contains(id)){
            throw Exception("Update ID already present")
        }

        val device: BluetoothDevice = bluetoothAdapter.getRemoteDevice(macAddress)
        val updateFileUri = Uri.parse(updateFileUriString)

        val upgrade = DeviceUpgrade(id, device, reactContext, updateFileUri, updateOptions, this)
        this.upgrades[id] = upgrade
    }

    @ReactMethod
    fun runUpgrade(id: String, promise: Promise) {
        if (!upgrades.contains(id)){
            promise.reject(Exception("update ID not present"))
        }

        upgrades[id]!!.startUpgrade(promise)
    }

    @ReactMethod
    fun cancelUpgrade(id: String) {
        if (!upgrades.contains(id)){
            Log.w(this.TAG,"can't cancel update ID ($id} not present")
            return
        }

        upgrades[id]!!.cancel()
    }

    @ReactMethod
    fun destroyUpgrade(id: String) {
        if (!upgrades.contains(id)){
            Log.w(this.TAG,"can't destroy update ID ($id} not present")
            return
        }

        upgrades[id]!!.cancel()
        upgrades.remove(id)
    }

    fun updateProgressCB(progress: WritableMap?) {
        reactContext
                .getJSModule(RCTDeviceEventEmitter::class.java)
                .emit("uploadProgress", progress)
    }

    fun upgradeStateCB(state: WritableMap?) {
        reactContext
                .getJSModule(RCTDeviceEventEmitter::class.java)
                .emit("upgradeStateChanged", state)
    }

    @ReactMethod
    fun addListener(eventName: String) {
        // Keep: Required for RN built in Event Emitter Calls.
    }

    @ReactMethod
    fun removeListeners(count: Integer) {
        // Keep: Required for RN built in Event Emitter Calls.
    }
}
